src/coreclr/src/gc/gc.cpp

```diff
@@ -37770,32 +37558,26 @@ GCHeap::GarbageCollect (int generation, bool low_memory_p, int mode)
         else
         {
             BOOL should_collect = FALSE;
-            BOOL should_check_loh = (generation == max_generation);
+            BOOL should_check_uoh = (generation == max_generation);
 #ifdef MULTIPLE_HEAPS
             for (int i = 0; i < gc_heap::n_heaps; i++)
             {
                 dynamic_data* dd1 = gc_heap::g_heaps [i]->dynamic_data_of (generation);
-                dynamic_data* dd2 = (should_check_loh ?
-                                     (gc_heap::g_heaps [i]->dynamic_data_of (max_generation + 1)) :
-                                     0);
-
-                if (should_collect_optimized (dd1, low_memory_p))
+                if (!should_collect && should_check_uoh)
```

Comment:
you still need to call should_collect_optimized on max_generation always; you just also need to check uoh if should_check_uoh is true (and should_collect_optimized on max_generation is false).

also if it's changed to uoh it should really check all uoh generations